server/server/main.py

```diff
@@ -0,0 +1,202 @@
+    # server to perform other tasks such as sending notifications
+    # TODO: convert pipeline.run to an async function
+    await asyncio.to_thread(pipeline.run)
+
+def unique_id():
+    return str(uuid.uuid4())
+
+@app.get("/sessions", response_model=List[SessionModel])
+async def get_sessions():
+    global context
+
+    return list(context.sessions.values())
+
+@app.post("/sessions", response_model=SessionModel)
+async def create_session():
+    global context
+
+    session_id = unique_id()
+    session = SessionModel(id=session_id)
+    context.sessions[session_id] = session
+
+    return session
+
+@app.get("/pipelines", response_model=List[SerializablePipelineModel])
+async def get_pipelines(
+    session_id: IdType = Query(..., alias='sessionId')
+):
+    global context
+
+    if session_id not in context.sessions:
+        raise Exception(f"Session {session_id} does not exist")
+
+    pipeline_ids = context.session_pipelines.get(session_id, set())
+
+    pipelines = list(map(lambda pipeline_id: context.pipelines[pipeline_id], pipeline_ids))
+
+    return pipelines
+
+
+@app.post("/pipelines", response_model=SerializablePipelineModel)
+async def create_pipeline(
+    background_tasks: BackgroundTasks,
+    session_id: IdType = Query(..., alias='sessionId'),
+    graph: PipelineJSON = Body(...)
+):
+    global context
+
+    if session_id not in context.sessions:
+        raise Exception(f"Session {session_id} does not exist")
+
+    pipeline_id = unique_id()
+    graph.id = pipeline_id
+    pipeline = deserialize_pipeline(graph.dict(by_alias=True), context.memory_client, context.registry)
+
+    def notify(message: NotificationMessage):
+        asyncio.run(notify_clients(message.dict(), session_id, context))
+
+    pipeline.observer = ServerPipelineObserver(notify)
+
+    model = PipelineModel(id=pipeline_id, graph=graph, pipeline=pipeline)
+
+    pipeline_ids = context.session_pipelines.setdefault(session_id, set())
+    pipeline_ids.add(model.id)
+
+    context.pipelines[model.id] = model
+
+    message = pipeline_created(model)
+
+    background_tasks.add_task(notify_clients, message.dict(), session_id, context)
+    background_tasks.add_task(run_pipeline, session_id, model.id, context)
+
+    return model
+
+@app.websocket("/ws")
+async def create_websocket(
+    socket: WebSocket,
+    session_id: IdType = Query(..., alias='sessionId')
+):
+    global context
+
+    await socket.accept()
+
+    websocket = WebsocketModel(id=unique_id(), socket=socket)
+
+    websocket_ids = context.session_websockets.setdefault(session_id, set())
+    websocket_ids.add(websocket.id)
+
+    context.websockets[websocket.id] = websocket
+
+    try:
+        while True:
+            _ = await socket.receive_json()
+    except WebSocketDisconnect:
+        websocket_ids.remove(websocket.id)
+
+@app.get("/")
+async def get_index():
+    index_file = f"{os.path.join(os.path.dirname(__file__), 'index.html')}"
```

Comment:
I think this is a little easier to do with `pathlib`'s `with_name()` function. For instance:

```python
>>> from pathlib import Path
>>> Path('/path/to/file.py').with_name('index.html')
PosixPath('/path/to/index.html')
```

Same for [here](https://github.com/OpenChemistry/oremda/pull/20/files#diff-b7e71dbe9962684ab6f05be10a8404725062a2a18836cba09decea66fc2aa169R200).